Classes/Private/Manager/Audio/OCTAudioEngine.m

```diff
@@ -253,13 +256,28 @@ static OSStatus inputRenderCallBack(void *inRefCon,
                                        inBusNumber,
                                        inNumberFrames,
                                        &bufferList);
-    NSError *error;
-    [engine.toxav sendAudioFrame:bufferList.mBuffers[0].mData
-                     sampleCount:inNumberFrames
-                        channels:kNumberOfChannels
-                      sampleRate:[engine currentAudioSampleRate] / 1000
-                        toFriend:engine.friendNumber
-                           error:&error];
+
+    TPCircularBufferProduceBytes(&engine->_outgoingBuffer,
+                                 bufferList.mBuffers[0].mData,
+                                 bufferList.mBuffers[0].mDataByteSize);
+
+    int32_t availableBytesToConsume;
+    void *tail = TPCircularBufferTail(&engine->_outgoingBuffer, &availableBytesToConsume);
+    int32_t minimalBytesToConsume = kSampleCount * kNumberOfChannels * sizeof(SInt16);
+
+    int32_t cyclesToConsume = availableBytesToConsume / minimalBytesToConsume;
+
+    for (int32_t i = 0; i < cyclesToConsume; i++) {
+        NSError *error;
+        [engine.toxav sendAudioFrame:tail
+                         sampleCount:kSampleCount
+                            channels:kNumberOfChannels
+                          sampleRate:engine.currentAudioSampleRate
+                            toFriend:engine.friendNumber
+                               error:&error];
+        TPCircularBufferConsume(&engine->_outgoingBuffer, minimalBytesToConsume);
+        tail = TPCircularBufferTail(&engine->_outgoingBuffer, &availableBytesToConsume);
```

Comment:
This naming is confusing. You are using `outgoingBuffer` for `inputRenderCallback` and `incomingBuffer` for `outputRenderCallback`. Maybe rename buffers to `inputBuffer` and `outputBuffer`?